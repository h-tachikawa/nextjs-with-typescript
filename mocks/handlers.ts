import { graphql } from "msw";
import type { GetCountries } from "../pages/__generated__/GetCountries";

export const handlers = [
  graphql.query<GetCountries>("GetCountries", (req, res, ctx) => {
    return res(
      ctx.data({
        countries: [
          {
            __typename: "Country",
            code: "JP",
            name: "Japan",
            emoji: "ðŸ‡¯ðŸ‡µ",
          },
          {
            __typename: "Country",
            code: "US",
            name: "United States",
            emoji: "ðŸ‡ºðŸ‡¸",
          },
        ],
      })
    );
  }),
];
