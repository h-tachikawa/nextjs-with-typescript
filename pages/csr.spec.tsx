import Csr from "./csr";
import { render, act } from "@testing-library/react";
import { graphql, GraphQLHandler, GraphQLRequest } from "msw";
import { sleep } from "../lib/sleep";
import client from "../apollo-client";
import { ApolloProvider } from "@apollo/client";
import { server } from "../mocks/server";
import React from "react";
import { GetCountries } from "./__generated__/GetCountries";

const useRouter = jest.spyOn(require("next/router"), "useRouter");
useRouter.mockImplementation(() => ({
  pathname: "/",
}));

jest.mock("next/link", () => {
  // @ts-ignore
  return ({ children }) => {
    return children;
  };
});

export const testRenderer =
  (children: React.ReactNode) =>
  (responseOverride?: GraphQLHandler<GraphQLRequest<never>>) => {
    if (responseOverride) {
      server.use(responseOverride);
    }
    return render(<ApolloProvider client={client}>{children}</ApolloProvider>);
  };

describe("Csr Component", () => {
  beforeEach(() => client.cache.reset());

  describe("ordinary", () => {
    it("renders without error", async () => {
      const render = testRenderer(<Csr />);
      const { asFragment } = render();

      await act(async () => {
        await sleep(10);
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("when empty response", () => {
    it("renders without error", async () => {
      const render = testRenderer(<Csr />);
      const { asFragment } = render(
        graphql.query<GetCountries>("GetCountries", (req, res, ctx) => {
          return res(
            ctx.data({
              countries: [],
            })
          );
        })
      );

      await act(async () => {
        await sleep(10);
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
