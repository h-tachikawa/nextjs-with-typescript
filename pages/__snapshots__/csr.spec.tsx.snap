// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Csr Component ordinary renders without error 1`] = `
<DocumentFragment>
  <div>
    <p>
      JP - ðŸ‡¯ðŸ‡µ
    </p>
    <p>
      US - ðŸ‡ºðŸ‡¸
    </p>
    <div
      class="MuiBox-root css-1274r9c"
    >
      <a
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  active css-nx6gxd-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
      >
        Go to the home page
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Csr Component when empty response renders without error 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root css-1274r9c"
    >
      <a
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  active css-nx6gxd-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
      >
        Go to the home page
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </a>
    </div>
  </div>
</DocumentFragment>
`;
